File: frontend/app/src/types/education.ts
/**
 * Normalized education API contracts shared by the admin/teacher/student
 * workspaces. These mirror the FastAPI response shapes in backend/api/{courses,
 * assignments,submissions,grading_runs,results,admin}.ts; the frontend branches
 * on the stable status unions (never on natural-language strings).
 */

export type AssignmentStatus = "draft" | "ready" | "published" | "closed" | "archived";

export type GradingRunStatus =
  | "queued"
  | "running"
  | "completed"
  | "partial_failed"
  | "failed"
  | "cancelled";

export type GradeResultStatus = "graded" | "failed" | "needs_review";

export interface Course {
  id: string;
  name: string;
  code: string;
  description: string;
  teacher_id: string;
  student_ids: string[];
  student_count: number;
  created_at: number;
  updated_at: number;
}

export interface Assignment {
  id: string;
  course_id: string;
  teacher_id: string;
  name: string;
  description: string;
  status: AssignmentStatus;
  due_at: number | null;
  created_at: number;
  updated_at: number;
  published_at: number | null;
  version: number;
  question_count: number;
}

export interface Question {
  id: string;
  assignment_id: string;
  q_id: string;
  order_index: number;
  number: string;
  type: string;
  stem: string;
  criterion: string;
  max_score: number;
  reference_answer: string | null;
  test_cases: unknown[] | null;
  source: Record<string, unknown> | null;
  version: number;
  created_at: number;
  updated_at: number;
}

export interface SubmissionSummary {
  id: string;
  assignment_id: string;
  student_id: string;
  current_revision_id: string | null;
  current_revision_number: number | null;
  created_at: number;
  updated_at: number;
}

export interface SubmissionAnswer {
  id: string;
  revision_id: string;
  question_id: string;
  q_id: string;
  number: string;
  type: string;
  content: string;
  flag: string[];
}

export interface SubmissionRevision {
  id: string;
  submission_id: string;
  revision_number: number;
  source: "online" | "teacher_import";
  file_name: string;
  created_at: number;
  answers: SubmissionAnswer[];
}

export interface GradingRun {
  id: string;
  assignment_id: string;
  teacher_id: string;
  status: GradingRunStatus;
  lease_owner: string | null;
  lease_expiry: number | null;
  last_heartbeat_at: number | null;
  total_submissions: number;
  completed_submissions: number;
  failed_submissions: number;
  error_message: string | null;
  created_at: number;
  started_at: number | null;
  completed_at: number | null;
  released_at: number | null;
}

export interface GradeResult {
  id: string;
  grading_run_id: string;
  question_id: string;
  q_id: string;
  student_id: string;
  ai_score: number | null;
  ai_max_score: number;
  ai_comment: string;
  result_status: GradeResultStatus;
  requires_review: boolean;
  review_reason: string | null;
  score: number | null;
  teacher_comment: string;
  effective_score?: number | null;
  effective_comment?: string;
  teacher_review?: {
    id: string;
    grade_result_id: string;
    teacher_id: string;
    previous_score: number | null;
    previous_comment: string | null;
    new_score: number;
    new_comment: string;
    created_at: number;
  } | null;
}

export interface AdminUser {
  id: string;
  username: string;
  email: string;
  role: UserRole;
  is_active: boolean;
  created_at: number;
}

export interface Invite {
  code: string;
  email: string | null;
  role: UserRole;
  course_id: string | null;
  created_at: number;
  expires_at: number;
  used_at: number | null;
  used_by: string | null;
}

/** Stable error envelope: every normalized router returns this shape on failure. */
export interface DomainErrorResponse {
  error: {
    code: string;
    message: string;
  };
}

import type { UserRole } from "./auth";
export type { UserRole };

File: backend/domain/education.py
"""Normalized education domain enums and immutable DTOs.

These types are the single business vocabulary shared by repositories,
services, and the API layer. ORM records (backend.db.models) persist them and
FastAPI request/response models mirror them, but neither owns the vocabulary —
this module does. Keeping the enums here (rather than on the ORM records)
means a status transition rule is enforced by code that imports this module
regardless of which table a row lives in.
"""
from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field, model_validator


class Role(str, Enum):
    ADMIN = "admin"
    TEACHER = "teacher"
    STUDENT = "student"


class AssignmentStatus(str, Enum):
    DRAFT = "draft"
    READY = "ready"
    PUBLISHED = "published"
    CLOSED = "closed"
    ARCHIVED = "archived"


# Assignments may be edited while in these states; publishing freezes the
# question set. Status transitions are validated by the assignment service
# against this set rather than by ad-hoc checks in endpoints.
EDITABLE_ASSIGNMENT_STATUSES = frozenset(
    {AssignmentStatus.DRAFT.value, AssignmentStatus.READY.value}
)


class GradingRunStatus(str, Enum):
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    PARTIAL_FAILED = "partial_failed"
    FAILED = "failed"
    CANCELLED = "cancelled"


# A run is "active" (and thus blocks a new run for the same assignment) while
# in one of these states. The DB partial unique index uses the same set.
ACTIVE_GRADING_RUN_STATUSES = frozenset(
    {GradingRunStatus.QUEUED.value, GradingRunStatus.RUNNING.value}
)


class GradeResultStatus(str, Enum):
    GRADED = "graded"
    FAILED = "failed"
    NEEDS_REVIEW = "needs_review"


# Results in these states are excluded from totals and from release; they route
# to the teacher review queue instead of being shown as a real score.
NON_GRADED_RESULT_STATUSES = frozenset(
    {GradeResultStatus.FAILED.value, GradeResultStatus.NEEDS_REVIEW.value}
)


class SubmissionRevisionSource(str, Enum):
    ONLINE = "online"
    TEACHER_IMPORT = "teacher_import"


# ─── Immutable DTOs (returned by repositories, never the ORM record) ──────────


class CourseDTO(BaseModel):
    id: str
    name: str
    code: str = ""
    description: str = ""
    teacher_id: str
    created_at: float
    updated_at: float
    student_ids: list[str] = Field(default_factory=list)


class EnrollmentDTO(BaseModel):
    course_id: str
    student_id: str
    enrolled_at: float


class QuestionDTO(BaseModel):
    id: str
    assignment_id: str
    q_id: str
    order_index: int
    number: str = ""
    type: str
    stem: str = ""
    criterion: str = ""
    max_score: float = 10.0
    reference_answer: str | None = None
    test_cases: list[dict[str, Any]] | None = None
    source: dict[str, Any] | None = None
    version: int = 1
    created_at: float
    updated_at: float


class AssignmentDTO(BaseModel):
    id: str
    course_id: str
    teacher_id: str
    name: str
    description: str = ""
    status: str = AssignmentStatus.DRAFT.value
    due_at: float | None = None
    created_at: float
    updated_at: float
    published_at: float | None = None
    version: int = 1
    question_count: int = 0


class SubmissionDTO(BaseModel):
    id: str
    assignment_id: str
    student_id: str
    current_revision_id: str | None = None
    current_revision_number: int | None = None
    created_at: float
    updated_at: float


class SubmissionRevisionDTO(BaseModel):
    id: str
    submission_id: str
    revision_number: int
    source: str = SubmissionRevisionSource.ONLINE.value
    file_name: str = ""
    created_at: float
    answers: list["SubmissionAnswerDTO"] = Field(default_factory=list)


class SubmissionAnswerDTO(BaseModel):
    id: str
    revision_id: str
    question_id: str
    q_id: str
    number: str = ""
    type: str = ""
    content: str = ""
    flag: list[str] = Field(default_factory=list)


class GradingRunDTO(BaseModel):
    id: str
    assignment_id: str
    teacher_id: str
    status: str = GradingRunStatus.QUEUED.value
    lease_owner: str | None = None
    lease_expiry: float | None = None
    last_heartbeat_at: float | None = None
    total_submissions: int = 0
    completed_submissions: int = 0
    failed_submissions: int = 0
    error_message: str | None = None
    created_at: float
    started_at: float | None = None
    completed_at: float | None = None
    released_at: float | None = None


class GradeResultDTO(BaseModel):
    id: str
    grading_run_id: str
    submission_revision_id: str
    question_id: str
    student_id: str
    q_id: str
    ai_score: float | None = None
    ai_max_score: float = 10.0
    ai_comment: str = ""
    ai_steps: list[dict[str, Any]] = Field(default_factory=list)
    ai_confidence: float | None = None
    ai_expert_results: list[dict[str, Any]] = Field(default_factory=list)
    ai_synthesis_method: str | None = None
    requires_review: bool = False
    review_reason: str | None = None
    initial_requires_review: bool = False
    initial_review_reason: str | None = None
    result_status: str = GradeResultStatus.GRADED.value
    effective_score: float | None = None
    effective_comment: str = ""
    teacher_review: dict[str, Any] | None = None
    created_at: float
    updated_at: float

    @model_validator(mode="after")
    def initialize_review_audit_facts(self) -> "GradeResultDTO":
        """Default omitted initial facts to the DTO's original live state.

        Adapters constructing a new result need not duplicate the same fields,
        while repository reads explicitly supply the durable audit values.
        """
        if "initial_requires_review" not in self.model_fields_set:
            self.initial_requires_review = self.requires_review
        if "initial_review_reason" not in self.model_fields_set:
            self.initial_review_reason = self.review_reason
        return self


class TeacherReviewDTO(BaseModel):
    id: str
    grade_result_id: str
    teacher_id: str
    previous_score: float | None = None
    previous_comment: str | None = None
    new_score: float
    new_comment: str = ""
    comment: str = ""
    confirmed: bool = False
    created_at: float


SubmissionRevisionDTO.model_rebuild()
